src/amqproxy/server.cr

```diff
@@ -73,10 +85,41 @@ module AMQProxy
           break
         end
       end
-      @log.info "Proxy stopping accepting connections"
+      wait_for_drain
+    end
+
+    # This function will be called by the close method and at the end of the listen loop, to make sure that
+    # the program exits and the close method only starts closing the socket etc. after all connections are drained.
+    # This means that it is called twice (on SIGTERM/SIGINT and always at the end),
+    # thus we make sure that the logs are only generated on SIGTERM/SIGINT via the `from_close_call` parameter
+    def wait_for_drain(from_close_call = false)
+      if !@graceful_shutdown
+        if from_close_call
+          @log.info "Skip waiting for open connections to be closed by remote"
+        end
+        return
+      end
+      if from_close_call
+        @log.info "Waiting for open connections to be closed by remote"
+      end
+      while @clients.size > 0
+        if from_close_call
+          @log.info "#{@clients.size} client(s) remaining"
+          @clients.each do |client|
+            @log.debug "- Client #{client.socket.as(TCPSocket).remote_address} since #{(client.lifetime.total_minutes.floor.to_i)}m #{client.lifetime.seconds}s"
+          end
+        end
+        sleep 20
```

Comment:
Quite long sleep, I think we should check more often (but maybe not log that often?)

```suggestion
        sleep 500.milliseconds
```

Reply:
Changed this